### PR TITLE
Assignment tab

### PR DESCRIPTION
**Pull request by WChrisK**

### CODE REVIEW

File: app/helpers/peer_review_helper.rb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+module PeerReviewHelper
+end
```
Comment (WChrisK) [inline app/helpers/peer_review_helper.rb]: Delete file?

File: app/models/assignment.rb [hunk 2]
```diff
@@ -914,6 +915,28 @@ def has_peer_review_assignment?
     not pr_assignment.nil?
   end
 
+  def create_peer_review_assignment_if_not_exist
+    if has_peer_review and Assignment.where(parent_assignment_id: id).empty?
+      peerreview_assignment = Assignment.new
+      peerreview_assignment.parent_assignment = self
+      peerreview_assignment.submission_rule = NoLateSubmissionRule.new
+      peerreview_assignment.assignment_stat = AssignmentStat.new
```
Comment (WChrisK) [inline app/models/assignment.rb]: There has to be a better way for the above two lines, copying it was triggering some bad errors... do we even care about submission rules or assignment stats for peer review?

File: app/models/assignment.rb [hunk 3]
```diff
@@ -914,6 +915,28 @@ def has_peer_review_assignment?
...
+      peerreview_assignment.token_period = 1
+      peerreview_assignment.unlimited_tokens = false
+      peerreview_assignment.short_identifier = short_identifier + 'PR'
```
Comment (WChrisK) [inline app/models/assignment.rb]: This is also a potential problem if a short identifier with this name exists. Before changing it, do you have any recommendations on what to do here?

File: app/helpers/peer_review_helper.rb (hunk 1 above)
Comment (david-yz-liu) [inline app/helpers/peer_review_helper.rb]: Sure.

File: app/models/assignment.rb (hunk 2 above)
Comment (david-yz-liu) [inline app/models/assignment.rb]: We do (they determine whether there can be late submissions and mark statistics). You've set the correct defaults here.

File: app/models/assignment.rb (hunk 3 above)
Comment (david-yz-liu) [inline app/models/assignment.rb]: No great ideas. Make the suffix `_pr`. Validation will fail on conflict, which is okay for now.

File: app/models/assignment.rb [hunk 4]
```diff
@@ -914,6 +915,28 @@ def has_peer_review_assignment?
...
+      peerreview_assignment.description = description
```
Comment (david-yz-liu) [inline app/models/assignment.rb]: Leave this blank by default.

File: app/controllers/peer_reviews_controller.rb [hunk 5]
```diff
@@ -1,75 +1,32 @@
 class PeerReviewsController < ApplicationController
-  before_action :set_peer_review, only: [:show, :edit, :update, :destroy]
```
Comment (david-yz-liu) [inline app/controllers/peer_reviews_controller.rb]: Okay so... you should not be changing this file in any way (yet). It has nothing to do with what you're doing here.

File: app/controllers/peer_reviews_controller.rb [hunk 6]
```diff
@@ -63,13 +63,13 @@ def destroy
   end
 
   private
-    # Use callbacks to share common setup or constraints between actions.
-    def set_peer_review
-      @peer_review = PeerReview.find(params[:id])
-    end
```
Comment (WChrisK) [inline app/controllers/peer_reviews_controller.rb]: Do you want this undone? Looks like rubymine auto-un-indented when converting back to the old version.

File: app/models/assignment.rb [hunk 7]
```diff
@@ -914,6 +915,28 @@ def has_peer_review_assignment?
...
+  def create_peer_review_assignment_if_not_exist
+    if has_peer_review and Assignment.where(parent_assignment_id: id).empty?
+      peerreview_assignment = Assignment.new
+      peerreview_assignment.parent_assignment = self
+      peerreview_assignment.submission_rule = NoLateSubmissionRule.new
+      peerreview_assignment.assignment_stat = AssignmentStat.new
+      peerreview_assignment.token_period = 1
+      peerreview_assignment.unlimited_tokens = false
+      peerreview_assignment.short_identifier = short_identifier + '_pr'
+      peerreview_assignment.description = ''
```
Comment (WChrisK) [inline app/models/assignment.rb]: Is this what you meant by leaving it blank? (or use `nil`? or something else?)

File: app/controllers/peer_reviews_controller.rb (hunk 6 above)
Comment (david-yz-liu) [inline app/controllers/peer_reviews_controller.rb]: Eh, I think this is actually more consistent with the style in other files. I wouldn't have put it in here, but you can leave it.

File: app/assets/stylesheets/peer_reviews.css [hunk 8]
```diff
@@ -0,0 +1 @@
+@import 'assignments';
```
Comment (david-yz-liu) [inline app/assets/stylesheets/peer_reviews.css]: Remove this file as well (doesn't come in yet). Though do remember this change for later.

File: app/models/assignment.rb (hunk 7 above)
Comment (david-yz-liu) [inline app/models/assignment.rb]: Yes, that's fine.

File: app/models/assignment.rb (hunk 2 above)
Comment (david-yz-liu) [inline app/models/assignment.rb]: Actually, can you please check to make sure that the records for NoLateSubmissionRule and AssignmentStat are properly saved in the database after this callback? Normally `new` does not save.

File: app/views/assignments/_peer_review_form.html.erb [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+                    t('assignment.short_identifier'),
+                    class: 'required' %>
+        <%= f.text_field :short_identifier,
+                         onchange: 'set_onbeforeunload(true);',
+                         size: 35,
+                         maxlength: 30,
+                         required: true %>
+      </div>
+
+      <div>
```
Comment (david-yz-liu) [inline app/views/assignments/_peer_review_form.html.erb]: Don't show this field.

File: app/views/assignments/_peer_review_form.html.erb [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+                         onchange: 'set_onbeforeunload(true);',
+                         value: @assignment.due_date.nil? ? '' :
+                             @assignment.due_date.strftime(
+                                 '%Y-%m-%d %l:%M %p'),
+                         size: 35,
+                         required: true,
+                         placeholder: t('date.formats.datetime_placeholder') %>
+      </div>
+
+      <div>
```
Comment (david-yz-liu) [inline app/views/assignments/_peer_review_form.html.erb]: Don't show this.

File: app/views/assignments/_peer_review_form.html.erb [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+        <%= f.check_box :allow_web_submits,
+                        onclick: 'set_onbeforeunload(true);' %>
+        <%= f.label :allow_web_submits,
+                    t('assignment.allow_web_submits') %>
+      </div>
+
+      <div>
```
Comment (david-yz-liu) [inline app/views/assignments/_peer_review_form.html.erb]: Don't show this.

File: app/views/assignments/_peer_review_form.html.erb [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+        <%= f.label :display_grader_names_to_students,
+                    t('assignment.display_grader_names_to_students') %>
+      </div>
+
+      <div>
+        <%= f.check_box :is_hidden,
+                        onclick: 'set_onbeforeunload(true);' %>
+        <%= f.label :is_hidden, t('assignment.is_hidden') %>
+      </div>
+      <div>
```
Comment (david-yz-liu) [inline app/views/assignments/_peer_review_form.html.erb]: Don't show this.

File: app/views/assignments/_peer_review_form.html.erb [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+        <%= f.check_box :enable_test,
+                        onclick: 'set_onbeforeunload(true);' %>
+        <%= f.label :enable_test,
+                    t('automated_tests.enable_test') %>
+      </div>
+      <div>
```
Comment (david-yz-liu) [inline app/views/assignments/_peer_review_form.html.erb]: Don't show this.

File: app/views/assignments/_peer_review_form.html.erb [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+
+      <% Assignment::MARKING_SCHEME_TYPE.each do |key, value| %>
+          <%= f.radio_button :marking_scheme_type, value %>
+          <%= f.label :marking_scheme_type,
+                      t("assignment.marking_scheme.#{value}"),
+                      value: value %>
+      <% end %>
+    </fieldset>
+
+    <%# Required files %>
```
Comment (david-yz-liu) [inline app/views/assignments/_peer_review_form.html.erb]: Don't show this.

File: app/views/assignments/_peer_review_form.html.erb [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+      </div>
+    </fieldset>
+
+    <%# Starter Code %>
+
+    <%= f.submit t(:submit),
+                 data: { disable_with: t('working') },
+                 onclick: 'set_onbeforeunload(false);' %>
+<% end %>
+<% if @assignment.can_upload_starter_code? && @assignment.id %>
```
Comment (david-yz-liu) [inline app/views/assignments/_peer_review_form.html.erb]: Delete this part of the code (it's not shown, but not useful).

File: app/views/assignments/_peer_review_form.html.erb [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+          <%= f.label :remark_message, t('assignment.remark.message') %>
+          <%= f.text_area :remark_message,
+                          rows: 8,
+                          cols: 65,
+                          onchange: 'set_onbeforeunload(true);' %>
+        </p>
+      </div>
+    </fieldset>
+
+    <%# Starter Code %>
```
Comment (david-yz-liu) [inline app/views/assignments/_peer_review_form.html.erb]: Delete this line as well.

File: app/views/layouts/_sub_menu.html.erb [hunk 17]
```diff
@@ -12,7 +12,8 @@
         controller.controller_name == 'annotation_categories' ||
         controller.controller_name == 'grade_entry_forms' ||
         controller.controller_name == 'automated_tests' ||
-        controller.controller_name == 'tags' %>
+        controller.controller_name == 'tags' ||
```
Comment (david-yz-liu) [inline app/views/layouts/_sub_menu.html.erb]: Please revert the changes in this file, and the index file you created below. (Again, this is later.)

File: app/models/assignment.rb (hunk 2 above)
Comment (WChrisK) [inline app/models/assignment.rb]: Interestingly it appears to get added to the database, so I assume save cascades?

**Pull request closed without merging**